// @machinomy/contracts/dist/build/wrappers/TestToken.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const truffle = require("truffle-contract");
var TestToken;
(function (TestToken) {
    TestToken.ARTIFACT = require('../contracts/TestToken.json');
    function isMintEvent(something) {
        return something.event === 'Mint';
    }
    TestToken.isMintEvent = isMintEvent;
    function isMintFinishedEvent(something) {
        return something.event === 'MintFinished';
    }
    TestToken.isMintFinishedEvent = isMintFinishedEvent;
    function isOwnershipTransferredEvent(something) {
        return something.event === 'OwnershipTransferred';
    }
    TestToken.isOwnershipTransferredEvent = isOwnershipTransferredEvent;
    function isApprovalEvent(something) {
        return something.event === 'Approval';
    }
    TestToken.isApprovalEvent = isApprovalEvent;
    function isTransferEvent(something) {
        return something.event === 'Transfer';
    }
    TestToken.isTransferEvent = isTransferEvent;
    function contract(provider, defaults) {
        let instance = truffle(TestToken.ARTIFACT);
        if (provider) {
            instance.setProvider(provider);
        }
        if (defaults) {
            instance.defaults(defaults);
        }
        return instance;
    }
    TestToken.contract = contract;
})(TestToken = exports.TestToken || (exports.TestToken = {}));
exports.default = TestToken;
