// react-router-transition/lib/AnimatedRoute.js
'use strict';

exports.__esModule = true;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _Route = require('react-router-dom/Route');

var _Route2 = _interopRequireDefault(_Route);

var _matchPath = require('react-router-dom/matchPath');

var _matchPath2 = _interopRequireDefault(_matchPath);

var _RouteTransition = require('./RouteTransition');

var _RouteTransition2 = _interopRequireDefault(_RouteTransition);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

/**
 * Here we only care about whether or not the pathname matches. If so,
 * we'll use the route's path as the key, otherwise we'll default it
 * to a string signifying no match.
 */
function getKey(_ref, path, exact) {
  var pathname = _ref.pathname;

  return (0, _matchPath2.default)(pathname, { exact: exact, path: path }) ? 'match' : 'no-match';
}

var AnimatedRoute = function AnimatedRoute(_ref2) {
  var component = _ref2.component,
      path = _ref2.path,
      exact = _ref2.exact,
      routeTransitionProps = _objectWithoutProperties(_ref2, ['component', 'path', 'exact']);

  return _react2.default.createElement(_Route2.default, {
    render: function render(_ref3) {
      var location = _ref3.location,
          match = _ref3.match;
      return _react2.default.createElement(
        _RouteTransition2.default,
        routeTransitionProps,
        _react2.default.createElement(_Route2.default, {
          key: getKey(location, path, exact),
          path: path,
          exact: exact,
          location: location,
          component: component
        })
      );
    }
  });
};

exports.default = AnimatedRoute;
